File: framework/languages/java/e2e/SpotService/Client/src/main/java/systems/zlink/e2e/spotservice/client/Scenarios/SmD5Scenario.java
package systems.zlink.e2e.spotservice.client.Scenarios;

import java.util.List;
import java.util.UUID;
import systems.zlink.e2e.spotservice.shared.Contracts;
import systems.zlink.stream.connector.ZLinkStreamConnector;

public final class SmD5Scenario extends SpotServiceScenarioContext {
    private SmD5Scenario(SpotServiceScenarioContext context) {
        super(context);
    }

    public static void run(SpotServiceScenarioContext context) {
        new SmD5Scenario(context).execute();
    }

    private void execute() {
        try {
            String actorId = "actor-sm-d5-notified-" + UUID.randomUUID().toString().replace("-", "");
            ZLinkStreamConnector connector = createStreamConnector(options().streamAEndpoint());
            Contracts.ActorProfile profile = new Contracts.ActorProfile("Disconnect", 5, List.of("disconnect"));
            try {
                connector.connect().submit().toCompletableFuture().join();
                Contracts.ActorAuthRes auth = connector
                    .request(new Contracts.ActorAuthReq(actorId, profile))
                    .submit(Contracts.ActorAuthRes.class).toCompletableFuture().join();
                ensure(actorId.equals(auth.actorId()), "SM-D5 auth actor mismatch");
                Contracts.ActorJoinRes joined = connector
                    .request(new Contracts.ActorJoinReq("room-a", profile, profile.tags()))
                    .metadata("actor-id", actorId)
                    .submit(Contracts.ActorJoinRes.class).toCompletableFuture().join();
                ensure(actorId.equals(joined.actorId()), "SM-D5 join actor mismatch");
            } finally {
                closeQuietly(connector);
            }

            Contracts.EvidenceSnapshot evidence = waitForPlayAEvidence(
                List.of("ActorUserDisconnected|play-a|room-a|" + actorId));
            ensure(evidence.entries().stream().anyMatch(entry ->
                    "ActorUserDisconnected".equals(entry.marker()) && actorId.equals(entry.value())),
                "SM-D5 expected selected actor disconnect callback evidence");

            System.out.println("scenario SM-D5 passed");
        } catch (Exception error) {
            throw new IllegalStateException("actor disconnect notify scenario failed", error);
        }

    }
}

File: framework/languages/java/e2e-kotlin/SpotService/Client/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/client/scenarios/SmD5Scenario.kt
package systems.zlink.e2e.kotlin.spotservice.client.scenarios

import systems.zlink.framework.kotlin.*

import java.util.UUID
import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.Env
import systems.zlink.e2e.kotlin.spotservice.client.support.createStreamConnector
import systems.zlink.e2e.kotlin.spotservice.client.support.ensure
import systems.zlink.e2e.kotlin.spotservice.client.support.postJson

internal object SmD5Scenario {
    suspend fun run() {
        val actorId = "actor-sm-d5-notified-" + UUID.randomUUID().toString().replace("-", "")
        val connector = createStreamConnector(Env.get("e2e.stream.a.endpoint"))
        try {
            val profile = Contracts.ActorProfile("Disconnect", 5, listOf("disconnect"))
            connector.connect().await()
            val auth = connector
                .request(Contracts.ActorAuthReq(actorId, profile))
                .awaitReply<Contracts.ActorAuthRes>()
            ensure(auth.actorId == actorId, "SM-D5 auth actor mismatch")
            val joined = connector
                .request(Contracts.ActorJoinReq("actor-room-a", profile, profile.tags))
                .metadata("actor-id", actorId)
                .awaitReply<Contracts.ActorJoinRes>()
            ensure(joined.actorId == actorId, "SM-D5 join actor mismatch")
            connector.close().await()
        } finally {
            try {
                connector.close().await()
            } catch (_: Exception) {
            }
        }

        val evidence = postJson(
            Env.get("e2e.http.session.endpoint"),
            "/evidence/wait",
            Contracts.EvidenceWaitReq(
                listOf(
                    "ActorUserDisconnected|session-a|actor-room-a|$actorId",
                ),
                10_000,
            ),
            Contracts.EvidenceSnapshot::class.java,
        )
        ensure(
            evidence.entries.any { it.marker == "ActorUserDisconnected" && it.value == actorId },
            "SM-D5 expected selected actor disconnect callback evidence",
        )

        println("scenario SM-D5 passed")
    }
}
